arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/NonEmptyListModule.kt

```diff
@@ -26,36 +31,37 @@ object NonEmptyListModule : SimpleModule(PackageVersion.VERSION) {
   }
 }
 
-private object NelSerializationConverter : StdConverter<Nel<*>, List<*>>() {
-  override fun convert(value: Nel<*>?): List<*>? = value?.all.orEmpty()
+class NonEmptyListSerializationConverter : StdConverter<NonEmptyList<*>, List<*>>() {
+  override fun convert(value: NonEmptyList<*>?): List<*>? = value?.all.orEmpty()
 }
 
-private class NelDeserializationConverter(private val elementType: JavaType) : StdConverter<List<Any?>, Nel<Any?>?>() {
+private class NonEmptyListDeserializationConverter(private val elementType: JavaType) :
+  StdConverter<List<Any?>, NonEmptyList<Any?>?>() {
 
-  override fun convert(value: List<*>?): Nel<*>? = value?.pipe { Nel.fromList(it).orNull() }
+  override fun convert(value: List<*>?): NonEmptyList<*>? =
+    value.toOption()
+      .map { NonEmptyList.fromList(it).getOrElse { throw IllegalArgumentException("list is empty") } }
+      .orNull()
 
   override fun getInputType(typeFactory: TypeFactory): JavaType =
     typeFactory.constructCollectionType(List::class.java, elementType)
 
   override fun getOutputType(typeFactory: TypeFactory): JavaType =
-    typeFactory.constructCollectionLikeType(Nel::class.java, elementType)
+    typeFactory.constructCollectionLikeType(NonEmptyList::class.java, elementType)
 }
 
-private object NonEmptyListDeserializerResolver : Deserializers.Base() {
-
-  private val NON_EMPTY_LIST = Nel::class.java
+object NonEmptyListDeserializerResolver : Deserializers.Base() {
 
-  override fun findBeanDeserializer(
-    type: JavaType,
+  override fun findCollectionDeserializer(
+    type: CollectionType,
     config: DeserializationConfig,
-    beanDesc: BeanDescription
-  ): JsonDeserializer<Nel<*>>? {
-    val rawClass = type.rawClass
-
-    return if (!NON_EMPTY_LIST.isAssignableFrom(rawClass)) {
-      null
+    beanDesc: BeanDescription,
+    elementTypeDeserializer: TypeDeserializer?,
+    elementDeserializer: JsonDeserializer<*>?
+  ): JsonDeserializer<NonEmptyList<*>>? =
+    if (NonEmptyList::class.java.isAssignableFrom(type.rawClass)) {
+      StdDelegatingDeserializer<NonEmptyList<*>>(NonEmptyListDeserializationConverter(type.bindings.getBoundType(0)))
```

Comment:
Same here, do we need to instantiate here or can we use an object or val singleton converter?

Reply:
thanks @raulraja, the `DeserializationConverter` requires a type parameter to determine input/output java type. i.e. it's more like `(T) -> Deserializer<Nel<T>>` in kotlin.